<!DOCTYPE html><html lang="ZH" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>系统发育分析.md | VEXYXING</title><meta name="author" content="VERYXING565"><meta name="copyright" content="VERYXING565"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="概述为了寻求各物种之间的系统发育关系，而去寻找各物种之间共有的同源单拷贝基因。我认为难点在于筛选数据，目前指定筛选的顺序为：  基因组：在一个目标分类单位中，如螯肢动物亚门，寻找其中所有可用的基因组数据，使用compleasm进行检验，保证使用的基因组具备高完整性和低冗余性，从而有较多的同源单拷贝基因可供选择。这里的阈值（S）我选择90%。即为第一轮筛选。  此时，得到一批基因组数据及其一组同源单">
<meta property="og:type" content="article">
<meta property="og:title" content="系统发育分析.md">
<meta property="og:url" content="http://example.com/2025/03/18/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/index.html">
<meta property="og:site_name" content="VEXYXING">
<meta property="og:description" content="概述为了寻求各物种之间的系统发育关系，而去寻找各物种之间共有的同源单拷贝基因。我认为难点在于筛选数据，目前指定筛选的顺序为：  基因组：在一个目标分类单位中，如螯肢动物亚门，寻找其中所有可用的基因组数据，使用compleasm进行检验，保证使用的基因组具备高完整性和低冗余性，从而有较多的同源单拷贝基因可供选择。这里的阈值（S）我选择90%。即为第一轮筛选。  此时，得到一批基因组数据及其一组同源单">
<meta property="og:locale">
<meta property="og:image" content="http://example.com/img/touxiang.jpg">
<meta property="article:published_time" content="2025-03-18T04:24:34.000Z">
<meta property="article:modified_time" content="2025-04-30T14:42:22.819Z">
<meta property="article:author" content="VERYXING565">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/img/touxiang.jpg"><script type="application/ld+json">{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "系统发育分析.md",
  "url": "http://example.com/2025/03/18/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/",
  "image": "http://example.com/img/touxiang.jpg",
  "datePublished": "2025-03-18T04:24:34.000Z",
  "dateModified": "2025-04-30T14:42:22.819Z",
  "author": [
    {
      "@type": "Person",
      "name": "VERYXING565",
      "url": "http://example.com/"
    }
  ]
}</script><link rel="shortcut icon" href="../../../../img/cat.jpg"><link rel="canonical" href="http://example.com/2025/03/18/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="../../../../css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/node-snackbar/dist/snackbar.min.css" media="print" onload="this.media='all'"><script>
    (() => {
      
    const saveToLocal = {
      set: (key, value, ttl) => {
        if (!ttl) return
        const expiry = Date.now() + ttl * 86400000
        localStorage.setItem(key, JSON.stringify({ value, expiry }))
      },
      get: key => {
        const itemStr = localStorage.getItem(key)
        if (!itemStr) return undefined
        const { value, expiry } = JSON.parse(itemStr)
        if (Date.now() > expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return value
      }
    }

    window.btf = {
      saveToLocal,
      getScript: (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        Object.entries(attr).forEach(([key, val]) => script.setAttribute(key, val))
        script.onload = script.onreadystatechange = () => {
          if (!script.readyState || /loaded|complete/.test(script.readyState)) resolve()
        }
        script.onerror = reject
        document.head.appendChild(script)
      }),
      getCSS: (url, id) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onload = link.onreadystatechange = () => {
          if (!link.readyState || /loaded|complete/.test(link.readyState)) resolve()
        }
        link.onerror = reject
        document.head.appendChild(link)
      }),
      addGlobalFn: (key, fn, name = false, parent = window) => {
        if (!false && key.startsWith('pjax')) return
        const globalFn = parent.globalFn || {}
        globalFn[key] = globalFn[key] || {}
        globalFn[key][name || Object.keys(globalFn[key]).length] = fn
        parent.globalFn = globalFn
      }
    }
  
      
      const activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      const activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }

      btf.activateDarkMode = activateDarkMode
      btf.activateLightMode = activateLightMode

      const theme = saveToLocal.get('theme')
    
          const mediaQueryDark = window.matchMedia('(prefers-color-scheme: dark)')
          const mediaQueryLight = window.matchMedia('(prefers-color-scheme: light)')
          
          if (theme === undefined) {
            if (mediaQueryLight.matches) activateLightMode()
            else if (mediaQueryDark.matches) activateDarkMode()
            else {
              const hour = new Date().getHours()
              const isNight = hour <= 6 || hour >= 18
              isNight ? activateDarkMode() : activateLightMode()
            }
            mediaQueryDark.addEventListener('change', () => {
              if (saveToLocal.get('theme') === undefined) {
                e.matches ? activateDarkMode() : activateLightMode()
              }
            })
          } else {
            theme === 'light' ? activateLightMode() : activateDarkMode()
          }
        
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        document.documentElement.classList.toggle('hide-aside', asideStatus === 'hide')
      }
    
      
    const detectApple = () => {
      if (/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)) {
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
  
    })()
  </script><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false,"highlightFullpage":false,"highlightMacStyle":false},
  copy: {
    success: 'Copy Successful',
    error: 'Copy Failed',
    noSupport: 'Browser Not Supported'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: 'Just now',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'medium_zoom',
  Snackbar: {"chs_to_cht":"You have switched to Traditional Chinese","cht_to_chs":"You have switched to Simplified Chinese","day_to_night":"You have switched to Dark Mode","night_to_day":"You have switched to Light Mode","bgLight":"#49b1f5","bgDark":"#2d3035","position":"top-center"},
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid/dist/infinitegrid.min.js',
    buttonText: 'Load More'
  },
  isPhotoFigcaption: true,
  islazyloadPlugin: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: true,
  },
  autoDarkmode: true
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '系统发育分析.md',
  isHighlightShrink: false,
  isToc: true,
  pageType: 'post'
}</script><meta name="generator" content="Hexo 7.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img text-center"><img src="../../../../img/touxiang.jpg" onerror="this.onerror=null;this.src='../../../../img/friend_404.gif'" alt="avatar"/></div><div class="site-data text-center"><a href="../../../../archives/"><div class="headline">Articles</div><div class="length-num">3</div></a><a href="../../../../tags/"><div class="headline">Tags</div><div class="length-num">0</div></a><a href="../../../../categories/"><div class="headline">Categories</div><div class="length-num">0</div></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="../../../../index.html"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="../../../../archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="../../../../comments/"><i class="fa-fw fas fa-envelope-open"></i><span> 留言板</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url(/img/top_image.png);"><nav id="nav"><span id="blog-info"><a class="nav-site-title" href="../../../../index.html"><span class="site-name">VEXYXING</span></a><a class="nav-page-title" href="../../../../index.html"><span class="site-name">系统发育分析.md</span></a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="../../../../index.html"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="../../../../archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="../../../../comments/"><i class="fa-fw fas fa-envelope-open"></i><span> 留言板</span></a></div></div><div id="toggle-menu"><span class="site-page"><i class="fas fa-bars fa-fw"></i></span></div></div></nav><div id="post-info"><h1 class="post-title">系统发育分析.md</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2025-03-18T04:24:34.000Z" title="Created 2025-03-18 12:24:34">2025-03-18</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2025-04-30T14:42:22.819Z" title="Updated 2025-04-30 22:42:22">2025-04-30</time></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title=""><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post Views:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="container post-content" id="article-container"><h1 id="概述"><a href="#概述" class="headerlink" title="概述"></a>概述</h1><p>为了寻求各物种之间的系统发育关系，而去寻找各物种之间共有的同源单拷贝基因。我认为难点在于筛选数据，目前指定筛选的顺序为：</p>
<ol>
<li><p>基因组：在一个目标分类单位中，如螯肢动物亚门，寻找其中所有可用的基因组数据，使用compleasm进行检验，保证使用的基因组具备高完整性和低冗余性，从而有较多的同源单拷贝基因可供选择。这里的阈值（S）我选择90%。即为第一轮筛选。</p>
<p> 此时，得到一批基因组数据及其一组同源单拷贝基因，同时，提取一个物种的这组同源单拷贝基因作为参考。</p>
</li>
<li><p>基因组二代测序数据：这是第二轮数据筛选，从NCBI上下载缺失数据的分类单位的二代测序数据，然后使用参考的单拷贝基因数据进行搜索该物种对应的单拷贝基因数据，软件使用的是aTRAM。</p>
</li>
<li><p>转录组测序数据：这是第三轮数据筛选，即使用Trinity组装转录本，提取单拷贝基因</p>
</li>
</ol>
<h1 id="1-一轮筛选：基因组"><a href="#1-一轮筛选：基因组" class="headerlink" title="1 一轮筛选：基因组"></a>1 一轮筛选：基因组</h1><h2 id="1-1-数据下载"><a href="#1-1-数据下载" class="headerlink" title="1.1 数据下载"></a>1.1 数据下载</h2><p>以螯肢动物亚门为例，在NCBI上手动搜索，将各个分类目中可用的基因组的GenBank号复制保存到一个txt文档中，各个GenBank以“\t”进行分割。</p>
<p>随后，使用<code>assembly_genome_download_link.py</code>组装ascp下载指令，输出为一个sh文件，该脚本运行需要联网。</p>
<p>最后，运行指令<code>bash download_genome.sh</code>，开始下载基因组。若有些基因组下载失败，就多运行几次这个指令。如果始终下载失败，就把对应的命令复制出来，单独运行。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">python assembly_genome_download_link.py genbank_id.txt download_genome.sh</span><br><span class="line">mkdir zhizhu</span><br><span class="line">cd zhizhu</span><br><span class="line">bash ../download_genome.sh</span><br><span class="line">gunzip *.gz</span><br></pre></td></tr></table></figure>
<h2 id="1-2-数据筛选和提取单拷贝基因"><a href="#1-2-数据筛选和提取单拷贝基因" class="headerlink" title="1.2 数据筛选和提取单拷贝基因"></a>1.2 数据筛选和提取单拷贝基因</h2><p>下载compleasm，使用其分析所有下载的基因组。然后使用<code>compleasm_out_analyse.py</code>统计结果，并将阈值以上的基因组之间共有的同源单拷贝基因组提取合并。我一般采用90，即BUSCO评估中S达到90以上。统计结果保存在<code>all_compleasm_results.csv</code>中，所有物种都有的单拷贝基因存储在<code>/merge/complete/</code>里面,<code>/merge/no_complete/</code>里面存放的单拷贝基因是部分物种缺失的，如果采用的建树方法是并联法，或许可以使用。</p>
<p>compleasm数据库比较难下载，可能需要科学上网。lineage的选择，可以去查看BUSCO官网（ <a target="_blank" rel="noopener" href="https://busco-data.ezlab.org/v5/data/information/lineages_list.2025-03-03.txt.tar.gz">https://busco-data.ezlab.org/v5/data/information/lineages_list.2025-03-03.txt.tar.gz</a> ）</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">cd ..</span><br><span class="line">wget https://github.com/huangnengCSU/compleasm/releases/download/v0.2.7/compleasm-0.2.7_x64-linux.tar.bz2</span><br><span class="line">tar -jxvf compleasm-0.2.7_x64-linux.tar.bz2</span><br><span class="line">export PATH=/path/to/compleasm_kit:$PATH</span><br><span class="line"></span><br><span class="line">compleasm.py download arthropoda_odb10</span><br><span class="line">cd zhizhu</span><br><span class="line">for i in *.fna</span><br><span class="line">do compleasm.py run -t -l arthropoda -a $i -o $i.out -L ../mb_downloads</span><br><span class="line">done</span><br><span class="line">python ../compleasm_out_analyse.py ./zhizhu 90 ../mb_downloads/arthropoda_odb10</span><br></pre></td></tr></table></figure>
<h2 id="1-3-整理"><a href="#1-3-整理" class="headerlink" title="1.3 整理"></a>1.3 整理</h2><p>把所有符合要求的物种的基因组文件和compleasm输出文件夹复制到另外一个新的文件夹。</p>
<h1 id="2-二轮筛选：转录组"><a href="#2-二轮筛选：转录组" class="headerlink" title="2 二轮筛选：转录组"></a>2 二轮筛选：转录组</h1><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> *.fna</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">name=<span class="variable">$&#123;i%.fna*&#125;</span></span><br><span class="line"><span class="built_in">echo</span> <span class="variable">$name</span></span><br><span class="line">python ../narmal_gff_compleasm.py ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna.out&quot;</span>/arthropoda_odb10/miniprot_output.gff ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna.out&quot;</span>/arthropoda_odb10/miniprot_output.normal.gff</span><br><span class="line">gffread -g ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna&quot;</span> ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna.out&quot;</span>/arthropoda_odb10/miniprot_output.normal.gff -x ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna.out&quot;</span>/arthropoda_odb10/nt.fa</span><br><span class="line">python ../add_target.py ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna.out&quot;</span>/arthropoda_odb10/miniprot_output.normal.gff ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna.out&quot;</span>/arthropoda_odb10/nt.fa ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna.out&quot;</span>/arthropoda_odb10/nt.added.fa</span><br><span class="line">python ../get_compleasm_nt_seq.py ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna.out&quot;</span>/arthropoda_odb10/nt.added.fa ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna.out&quot;</span>/arthropoda_odb10/gene_marker.fasta ./<span class="string">&quot;<span class="variable">$&#123;name&#125;</span>.fna.out&quot;</span>/arthropoda_odb10/compleasm_nt.fa</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>



<h1 id="流程中使用的脚本"><a href="#流程中使用的脚本" class="headerlink" title="流程中使用的脚本"></a>流程中使用的脚本</h1><h2 id="assembly-genome-download-link-py"><a href="#assembly-genome-download-link-py" class="headerlink" title="assembly_genome_download_link.py"></a>assembly_genome_download_link.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># assembly_genome_download_link.py</span></span><br><span class="line"><span class="comment"># Usage:python assembly_genome_download_link.py genbank_id.txt downloda_genome.sh</span></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> requests</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">read_txt</span>(<span class="params">path2txt</span>):</span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(path2txt, <span class="string">&#x27;r&#x27;</span>, encoding=<span class="string">&#x27;utf-8&#x27;</span>) <span class="keyword">as</span> file:</span><br><span class="line">        lines_uncleaned = file.readlines()</span><br><span class="line">    lines_uncleaned = [line.strip() <span class="keyword">for</span> line <span class="keyword">in</span> lines_uncleaned]</span><br><span class="line">    lines = []</span><br><span class="line">    <span class="keyword">for</span> line_2 <span class="keyword">in</span> lines_uncleaned:</span><br><span class="line">        <span class="keyword">if</span> <span class="string">&#x27;#&#x27;</span> <span class="keyword">not</span> <span class="keyword">in</span> line_2:</span><br><span class="line">            lines.append(line_2)</span><br><span class="line">    <span class="keyword">return</span> lines</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">url_prepare</span>(<span class="params">line</span>):</span><br><span class="line">    url_name = <span class="string">&#x27;https://www.ncbi.nlm.nih.gov/datasets/genome/&#x27;</span> + line</span><br><span class="line"></span><br><span class="line">    line_1 = line.split(<span class="string">&#x27;_&#x27;</span>)[<span class="number">0</span>]</span><br><span class="line">    line_2 = line.split(<span class="string">&#x27;_&#x27;</span>)[<span class="number">1</span>]</span><br><span class="line">    line_2_1 = line_2[:<span class="number">3</span>]</span><br><span class="line">    line_2_2 = line_2[<span class="number">3</span>:<span class="number">6</span>]</span><br><span class="line">    line_2_3 = line_2[<span class="number">6</span>:].split(<span class="string">&#x27;.&#x27;</span>)[<span class="number">0</span>]</span><br><span class="line">    url_fna = <span class="string">&#x27;https://ftp.ncbi.nlm.nih.gov/genomes/all/&#x27;</span> + line_1 + <span class="string">&quot;/&quot;</span> + line_2_1 + <span class="string">&quot;/&quot;</span> + line_2_2 + <span class="string">&quot;/&quot;</span> + line_2_3</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> url_name,url_fna</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_name</span>(<span class="params">url_name</span>):</span><br><span class="line">    response = requests.get(url_name)</span><br><span class="line">    html_content = response.text</span><br><span class="line">    match_name = <span class="built_in">str</span>(re.search(<span class="string">r&#x27;pageData\.speciesName = &quot;([^&quot;]+)&quot;&#x27;</span>, html_content)).split(<span class="string">&#x27;&quot;&#x27;</span>)[<span class="number">1</span>].replace(<span class="string">&#x27; &#x27;</span>,<span class="string">&#x27;_&#x27;</span>)</span><br><span class="line">    <span class="keyword">return</span> match_name</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_url</span>(<span class="params">url_fna,line</span>):</span><br><span class="line">    response = requests.get(url_fna)</span><br><span class="line">    html_content = response.text</span><br><span class="line">    ex1 = <span class="string">&#x27;&lt;a href=(.*?)/&lt;/a&gt;&#x27;</span></span><br><span class="line">    part_1 = <span class="built_in">str</span>(re.findall(ex1, html_content, re.S)).split(<span class="string">&#x27;&quot;&#x27;</span>)</span><br><span class="line">    aa = <span class="number">0</span></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> part_1:</span><br><span class="line">        <span class="keyword">if</span> line <span class="keyword">in</span> i <span class="keyword">and</span> <span class="string">&#x27;]&#x27;</span> <span class="keyword">not</span> <span class="keyword">in</span> i <span class="keyword">and</span> aa == <span class="number">0</span>:</span><br><span class="line">            url_fna_2 = url_fna + <span class="string">&#x27;/&#x27;</span> + i + i.split(<span class="string">&#x27;/&#x27;</span>)[<span class="number">0</span>] + <span class="string">&#x27;_genomic.fna.gz&#x27;</span></span><br><span class="line">            aa += <span class="number">1</span></span><br><span class="line">            <span class="keyword">return</span> url_fna_2</span><br><span class="line">        <span class="keyword">elif</span> line <span class="keyword">in</span> i <span class="keyword">and</span> <span class="string">&#x27;]&#x27;</span> <span class="keyword">not</span> <span class="keyword">in</span> i <span class="keyword">and</span> aa != <span class="number">0</span>:</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">&quot;error:&quot;</span>,line)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">write_downloads_commend</span>(<span class="params">name,url,outpath,count,line</span>):</span><br><span class="line">    commend_fna = url.replace(<span class="string">&quot;https://ftp.ncbi.nlm.nih.gov&quot;</span>,</span><br><span class="line">     <span class="string">&quot;ascp -v -k 1 -T -l 200m -i ~/miniconda3/envs/download/etc/asperaweb_id_dsa.openssh anonftp@ftp.ncbi.nlm.nih.gov:&quot;</span>)\</span><br><span class="line">              + <span class="string">&#x27; ./&#x27;</span> + name + <span class="string">&#x27;_&#x27;</span> + line + <span class="string">&#x27;.fna.gz&#x27;</span></span><br><span class="line">    <span class="comment"># commend_gff = commend_fna.replace(&#x27;fna&#x27;, &#x27;gff&#x27;)</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">    <span class="keyword">if</span> count == <span class="number">0</span>:</span><br><span class="line">        <span class="keyword">with</span> <span class="built_in">open</span>(outpath, <span class="string">&quot;w&quot;</span>, encoding=<span class="string">&quot;utf-8&quot;</span>) <span class="keyword">as</span> file:</span><br><span class="line">            file.write(commend_fna.replace(<span class="string">&quot;&gt;&quot;</span>, <span class="string">&quot;&quot;</span>) + <span class="string">&quot;\n&quot;</span>)</span><br><span class="line">            <span class="comment"># file.write(commend_gff.replace(&quot;&gt;&quot;, &quot;&quot;) + &quot;\n&quot;)</span></span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="keyword">with</span> <span class="built_in">open</span>(outpath, <span class="string">&quot;a&quot;</span>, encoding=<span class="string">&quot;utf-8&quot;</span>) <span class="keyword">as</span> file:</span><br><span class="line">            file.write(commend_fna.replace(<span class="string">&quot;&gt;&quot;</span>, <span class="string">&quot;&quot;</span>) + <span class="string">&quot;\n&quot;</span>)</span><br><span class="line">            <span class="comment"># file.write(commend_gff.replace(&quot;&gt;&quot;, &quot;&quot;) + &quot;\n&quot;)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">    path2txt = sys.argv[<span class="number">1</span>]</span><br><span class="line">    outpath = sys.argv[<span class="number">2</span>]</span><br><span class="line">    lines = read_txt(path2txt=path2txt)</span><br><span class="line">    number = <span class="built_in">len</span>(lines)</span><br><span class="line">    count = <span class="number">0</span></span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> lines:</span><br><span class="line">        <span class="keyword">try</span>:</span><br><span class="line">            <span class="keyword">if</span> line == <span class="string">&quot;\t&quot;</span></span><br><span class="line">            url_name,url_fna = url_prepare(line)</span><br><span class="line">            name = get_name(url_name)</span><br><span class="line">            url = get_url(url_fna,line)</span><br><span class="line">            write_downloads_commend(name,url,outpath,count,line)</span><br><span class="line">            count += <span class="number">1</span></span><br><span class="line">            <span class="built_in">print</span>(<span class="built_in">str</span>(count) + <span class="string">&quot;/&quot;</span> + <span class="built_in">str</span>(number))</span><br><span class="line">        <span class="keyword">except</span> requests.RequestException <span class="keyword">as</span> e:</span><br><span class="line">            <span class="built_in">print</span>(line)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&quot;finished&quot;</span>)</span><br></pre></td></tr></table></figure>
<hr>
<h2 id="compleasm-out-analyse-py"><a href="#compleasm-out-analyse-py" class="headerlink" title="compleasm_out_analyse.py"></a>compleasm_out_analyse.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># compleasm_out_analyse.py</span></span><br><span class="line"><span class="comment"># Usage：python compleasm_out_analyse.py &lt;compleasm_out_path&gt; &lt;threshold&gt; &lt;compleasm_lineage_path]&gt;</span></span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line"><span class="keyword">import</span> glob</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">calculate</span>(<span class="params">path</span>):</span><br><span class="line">    results = []</span><br><span class="line">    folders = [f <span class="keyword">for</span> f <span class="keyword">in</span> os.listdir(path) <span class="keyword">if</span> os.path.isdir(os.path.join(path, f))]</span><br><span class="line">    <span class="keyword">for</span> folder <span class="keyword">in</span> folders:</span><br><span class="line">        <span class="keyword">if</span> folder == <span class="string">&quot;rename_single_gene&quot;</span>:</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line">        <span class="keyword">if</span> folder == <span class="string">&quot;merge&quot;</span>:</span><br><span class="line">            <span class="keyword">continue</span>    </span><br><span class="line">        name = folder.split(<span class="string">&quot;_&quot;</span>)[<span class="number">0</span>] + <span class="string">&quot;_&quot;</span> + folder.split(<span class="string">&quot;_&quot;</span>)[<span class="number">1</span>]</span><br><span class="line">        ncbi_id = folder.replace(name, <span class="string">&quot;&quot;</span>).replace(<span class="string">&quot;.fna.out&quot;</span>, <span class="string">&quot;&quot;</span>)</span><br><span class="line">        tmp_path = os.path.join(path, folder)</span><br><span class="line">        summary_path = os.path.join(tmp_path, <span class="string">&quot;summary.txt&quot;</span>)</span><br><span class="line">        <span class="keyword">with</span> <span class="built_in">open</span>(summary_path, <span class="string">&#x27;r&#x27;</span>) <span class="keyword">as</span> f:</span><br><span class="line">            <span class="keyword">for</span> line <span class="keyword">in</span> f:</span><br><span class="line">                <span class="keyword">if</span> line.startswith(<span class="string">&quot;#&quot;</span>):</span><br><span class="line">                    <span class="keyword">continue</span></span><br><span class="line">                <span class="keyword">elif</span> <span class="string">&quot;S&quot;</span> == line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">0</span>]:</span><br><span class="line">                    S_Percentage = line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>].split(<span class="string">&quot;,&quot;</span>)[<span class="number">0</span>]</span><br><span class="line">                    S_Number = line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>].split(<span class="string">&quot;,&quot;</span>)[<span class="number">1</span>]</span><br><span class="line">                <span class="keyword">elif</span> <span class="string">&quot;D&quot;</span> == line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">0</span>]:</span><br><span class="line">                    D_Percentage = line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>].split(<span class="string">&quot;,&quot;</span>)[<span class="number">0</span>]</span><br><span class="line">                    D_Number = line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>].split(<span class="string">&quot;,&quot;</span>)[<span class="number">1</span>]</span><br><span class="line">                <span class="keyword">elif</span> <span class="string">&quot;I&quot;</span> == line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">0</span>]:</span><br><span class="line">                    I_Percentage = line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>].split(<span class="string">&quot;,&quot;</span>)[<span class="number">0</span>]</span><br><span class="line">                    I_Number = line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>].split(<span class="string">&quot;,&quot;</span>)[<span class="number">1</span>]</span><br><span class="line">                <span class="keyword">elif</span> <span class="string">&quot;M&quot;</span> == line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">0</span>]:</span><br><span class="line">                    M_Percentage = line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>].split(<span class="string">&quot;,&quot;</span>)[<span class="number">0</span>]</span><br><span class="line">                    M_Number = line.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>].split(<span class="string">&quot;,&quot;</span>)[<span class="number">1</span>]</span><br><span class="line"></span><br><span class="line">        results.append(&#123;</span><br><span class="line">            <span class="string">&quot;name&quot;</span>: name,</span><br><span class="line">            <span class="string">&quot;ncbi_id&quot;</span>: ncbi_id,</span><br><span class="line">            <span class="string">&quot;S_Percentage&quot;</span>: S_Percentage,</span><br><span class="line">            <span class="string">&quot;S_Number&quot;</span>: S_Number,</span><br><span class="line">            <span class="string">&quot;D_Percentage&quot;</span>: D_Percentage,</span><br><span class="line">            <span class="string">&quot;D_Number&quot;</span>: D_Number,</span><br><span class="line">            <span class="string">&quot;I_Percentage&quot;</span>: I_Percentage,</span><br><span class="line">            <span class="string">&quot;I_Number&quot;</span>: I_Number,</span><br><span class="line">            <span class="string">&quot;M_Percentage&quot;</span>: M_Percentage,</span><br><span class="line">            <span class="string">&quot;M_Number&quot;</span>: M_Number</span><br><span class="line">        &#125;)</span><br><span class="line">    summary_df = pd.DataFrame(results)</span><br><span class="line">    output_csv = os.path.join(path, <span class="string">&quot;all_compleasm_results.csv&quot;</span>)</span><br><span class="line">    summary_df.to_csv(output_csv, index=<span class="literal">False</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;step 1 ok!!! 统计完成！结果已保存到 <span class="subst">&#123;output_csv&#125;</span>&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">seq_rename</span>(<span class="params">path, basedata_path</span>):</span><br><span class="line">    folders = [f <span class="keyword">for</span> f <span class="keyword">in</span> os.listdir(path) <span class="keyword">if</span> os.path.isdir(os.path.join(path, f))]</span><br><span class="line">    out_path = os.path.join(path, <span class="string">&quot;rename_single_gene&quot;</span>)</span><br><span class="line">    <span class="built_in">print</span>(out_path)</span><br><span class="line">    os.makedirs(out_path, exist_ok=<span class="literal">True</span>)</span><br><span class="line">    mid_name = os.path.basename(basedata_path)</span><br><span class="line">    count_2222 = <span class="number">0</span></span><br><span class="line">    <span class="keyword">for</span> folder <span class="keyword">in</span> folders:</span><br><span class="line">        <span class="keyword">if</span> folder == <span class="string">&quot;rename_single_gene&quot;</span>:</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line">        <span class="keyword">if</span> folder == <span class="string">&quot;merge&quot;</span>:</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line">        suffix = folder.split(<span class="string">&quot;.fna&quot;</span>)[<span class="number">0</span>] + <span class="string">&quot;.fna.out/&quot;</span> + <span class="string">&quot;arthropoda_odb10/&quot;</span> + <span class="string">&quot;gene_marker.fasta&quot;</span></span><br><span class="line">        base_name = folder.replace(<span class="string">&quot;.fna.out&quot;</span>, <span class="string">&quot;&quot;</span>)</span><br><span class="line">        gene_marker_path = os.path.join(path, suffix)</span><br><span class="line">        records = SeqIO.parse(gene_marker_path, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">        new_records = []</span><br><span class="line">        <span class="comment">#count_2222 += 1</span></span><br><span class="line">        <span class="keyword">for</span> record <span class="keyword">in</span> records:</span><br><span class="line">            new_record = record</span><br><span class="line">            new_record.<span class="built_in">id</span> = base_name + <span class="string">&quot;-&quot;</span> + record.<span class="built_in">id</span>.split(<span class="string">&quot;_&quot;</span>)[<span class="number">0</span>]</span><br><span class="line">            new_records.append(new_record)</span><br><span class="line">            <span class="comment"># print(record.id)</span></span><br><span class="line">        output_file = os.path.join(out_path, base_name + <span class="string">&quot;_gene_marker.fasta&quot;</span>)</span><br><span class="line">        <span class="comment">#print(output_file)</span></span><br><span class="line">        <span class="keyword">with</span> <span class="built_in">open</span>(output_file, <span class="string">&quot;w&quot;</span>) <span class="keyword">as</span> out_f:</span><br><span class="line">            SeqIO.write(new_records, out_f, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">    <span class="comment">#print(count_2222)</span></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&quot;step 2 ok!!!&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">Search_Common_Seq</span>(<span class="params">path, threshold, basedata_path</span>):</span><br><span class="line">    <span class="built_in">print</span></span><br><span class="line">    input_path = os.path.join(path, <span class="string">&quot;rename_single_gene&quot;</span>)</span><br><span class="line">    output_path = os.path.join(path, <span class="string">&quot;merge&quot;</span>)</span><br><span class="line">    os.makedirs(output_path, exist_ok=<span class="literal">True</span>)</span><br><span class="line">    threshold_out_path = os.path.join(output_path, <span class="string">&quot;complete&quot;</span>)</span><br><span class="line">    os.makedirs(threshold_out_path, exist_ok=<span class="literal">True</span>)</span><br><span class="line">    under_threshold_out_path = os.path.join(output_path, <span class="string">&quot;no_complete&quot;</span>)</span><br><span class="line">    os.makedirs(under_threshold_out_path, exist_ok=<span class="literal">True</span>)</span><br><span class="line">    all_records = []</span><br><span class="line">    threshold_count = <span class="number">0</span></span><br><span class="line">    tmp_hmm_path = os.path.join(basedata_path, <span class="string">&quot;hmms&quot;</span>)</span><br><span class="line">    hmm_files = glob.glob(os.path.join(tmp_hmm_path, <span class="string">&quot;*.hmm&quot;</span>))</span><br><span class="line">    fasta_files = glob.glob(os.path.join(input_path, <span class="string">&quot;*.fasta&quot;</span>))</span><br><span class="line">    genome_count = <span class="number">0</span></span><br><span class="line">    <span class="keyword">for</span> file <span class="keyword">in</span> fasta_files:</span><br><span class="line">        <span class="comment">#print(file)</span></span><br><span class="line">        tmp_records = SeqIO.parse(file, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">        </span><br><span class="line">        len_count = <span class="number">0</span></span><br><span class="line">        <span class="keyword">for</span> i <span class="keyword">in</span> tmp_records:</span><br><span class="line">            len_count += <span class="number">1</span></span><br><span class="line">        <span class="comment"># print(len_count, threshold * len(hmm_files) * 0.01)</span></span><br><span class="line">        <span class="comment">#print(threshold * len(hmm_files) * 0.01)</span></span><br><span class="line">        <span class="keyword">if</span> len_count &lt; threshold * <span class="built_in">len</span>(hmm_files) * <span class="number">0.01</span>:</span><br><span class="line">            </span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line">        genome_count += <span class="number">1</span></span><br><span class="line">        <span class="keyword">for</span> record <span class="keyword">in</span> SeqIO.parse(file, <span class="string">&quot;fasta&quot;</span>):</span><br><span class="line">            all_records.append(record)</span><br><span class="line">            <span class="comment">#print(genome_count)</span></span><br><span class="line">            <span class="comment">#print(record.id)</span></span><br><span class="line">        <span class="comment">#print(genome_count)</span></span><br><span class="line"></span><br><span class="line">    hmm_names = [os.path.basename(file) <span class="keyword">for</span> file <span class="keyword">in</span> hmm_files]</span><br><span class="line">    <span class="keyword">for</span> hmm <span class="keyword">in</span> hmm_names:</span><br><span class="line">        <span class="comment"># print(hmm)</span></span><br><span class="line">        hmm_records = []</span><br><span class="line">        hmm_base = hmm.split(<span class="string">&quot;.&quot;</span>)[<span class="number">0</span>]</span><br><span class="line">        <span class="comment">#print(hmm_base)</span></span><br><span class="line">        <span class="keyword">for</span> record <span class="keyword">in</span> all_records:</span><br><span class="line">            <span class="comment">#print(record.id)</span></span><br><span class="line">            <span class="keyword">if</span> hmm_base == record.<span class="built_in">id</span>.split(<span class="string">&quot;-&quot;</span>)[<span class="number">1</span>]:</span><br><span class="line">                <span class="comment">#print(record.id.split(&quot;-&quot;)[1])</span></span><br><span class="line">                <span class="comment"># print(hmm_base)</span></span><br><span class="line">                hmm_records.append(record)</span><br><span class="line">        </span><br><span class="line">        tmp_tmp_count = <span class="number">0</span></span><br><span class="line">        <span class="comment">#print(len(hmm_records))</span></span><br><span class="line">                </span><br><span class="line">        <span class="keyword">if</span> <span class="built_in">len</span>(hmm_records) == genome_count:</span><br><span class="line">            threshold_count += <span class="number">1</span></span><br><span class="line">            hmm_out_file = os.path.join(threshold_out_path, hmm)</span><br><span class="line">            <span class="keyword">with</span> <span class="built_in">open</span>(hmm_out_file, <span class="string">&quot;w&quot;</span>) <span class="keyword">as</span> out_f:</span><br><span class="line">                SeqIO.write(hmm_records, out_f, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            hmm_out_file = os.path.join(under_threshold_out_path, hmm)</span><br><span class="line">            <span class="keyword">with</span> <span class="built_in">open</span>(hmm_out_file, <span class="string">&quot;w&quot;</span>) <span class="keyword">as</span> out_f:</span><br><span class="line">                SeqIO.write(hmm_records, out_f, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&quot;step 3 ok!!!\nfinished&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">        compleasm_out_path = sys.argv[<span class="number">1</span>]</span><br><span class="line">        threshold = <span class="built_in">int</span>(sys.argv[<span class="number">2</span>])</span><br><span class="line">        compleasm_basedata_path = sys.argv[<span class="number">3</span>]</span><br><span class="line">        calculate(compleasm_out_path)</span><br><span class="line">        seq_rename(compleasm_out_path, compleasm_basedata_path)</span><br><span class="line">        Search_Common_Seq(compleasm_out_path, threshold, compleasm_basedata_path)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h2 id="normal-gff-of-compleasm-py"><a href="#normal-gff-of-compleasm-py" class="headerlink" title="normal_gff_of_compleasm.py"></a>normal_gff_of_compleasm.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># normal_gff_of_compleasm.py</span></span><br><span class="line"><span class="comment"># Usage：python normal_gff_of_compleasm.py &lt;input_gff&gt; &lt;output_gff&gt;</span></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">remove_comments</span>(<span class="params">input_gff, output_gff</span>):</span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(input_gff, <span class="string">&#x27;r&#x27;</span>) <span class="keyword">as</span> infile, <span class="built_in">open</span>(output_gff, <span class="string">&#x27;w&#x27;</span>) <span class="keyword">as</span> outfile:</span><br><span class="line">        <span class="keyword">for</span> line <span class="keyword">in</span> infile:</span><br><span class="line">            <span class="keyword">if</span> <span class="keyword">not</span> line.startswith(<span class="string">&#x27;#&#x27;</span>):</span><br><span class="line">                outfile.write(line)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">    input_gff = sys.argv[<span class="number">1</span>]</span><br><span class="line">    output_gff = sys.argv[<span class="number">2</span>]</span><br><span class="line">    remove_comments(input_gff, output_gff)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h2 id="add-description-py"><a href="#add-description-py" class="headerlink" title="add_description.py"></a>add_description.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># add_description.py</span></span><br><span class="line"><span class="comment"># Usage：python add_description.py &lt;normal_gff&gt; &lt;nt_file&gt; &lt;output_description_nt_fasta&gt;</span></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">from</span> collections <span class="keyword">import</span> defaultdict</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">parse_gff_target</span>(<span class="params">gff_file</span>):</span><br><span class="line">    id_to_target = <span class="built_in">dict</span>()</span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(gff_file, <span class="string">&#x27;r&#x27;</span>) <span class="keyword">as</span> f:</span><br><span class="line">        <span class="keyword">for</span> line <span class="keyword">in</span> f:</span><br><span class="line">            <span class="keyword">if</span> line.startswith(<span class="string">&#x27;#&#x27;</span>):</span><br><span class="line">                <span class="keyword">continue</span></span><br><span class="line">            parts = line.strip().split(<span class="string">&#x27;\t&#x27;</span>)</span><br><span class="line">            <span class="keyword">if</span> <span class="built_in">len</span>(parts) &lt; <span class="number">9</span>:</span><br><span class="line">                <span class="keyword">continue</span></span><br><span class="line">            attrs = parts[<span class="number">8</span>]</span><br><span class="line">            attr_dict = &#123;&#125;</span><br><span class="line">            <span class="keyword">for</span> attr <span class="keyword">in</span> attrs.split(<span class="string">&#x27;;&#x27;</span>):</span><br><span class="line">                <span class="keyword">if</span> <span class="string">&#x27;=&#x27;</span> <span class="keyword">in</span> attr:</span><br><span class="line">                    key, value = attr.split(<span class="string">&#x27;=&#x27;</span>, <span class="number">1</span>)</span><br><span class="line">                    attr_dict[key] = value</span><br><span class="line">            <span class="keyword">if</span> <span class="string">&#x27;ID&#x27;</span> <span class="keyword">in</span> attr_dict <span class="keyword">and</span> <span class="string">&#x27;Target&#x27;</span> <span class="keyword">in</span> attr_dict:</span><br><span class="line">                id_to_target[attr_dict[<span class="string">&#x27;ID&#x27;</span>]] = attr_dict[<span class="string">&#x27;Target&#x27;</span>]</span><br><span class="line">            <span class="keyword">elif</span> <span class="string">&#x27;Parent&#x27;</span> <span class="keyword">in</span> attr_dict <span class="keyword">and</span> <span class="string">&#x27;Target&#x27;</span> <span class="keyword">in</span> attr_dict:</span><br><span class="line">                id_to_target[attr_dict[<span class="string">&#x27;Parent&#x27;</span>]] = attr_dict[<span class="string">&#x27;Target&#x27;</span>]</span><br><span class="line">    <span class="keyword">return</span> id_to_target</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">add_target_to_fasta</span>(<span class="params">fasta_file, id_to_target, output_file</span>):</span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(fasta_file, <span class="string">&#x27;r&#x27;</span>) <span class="keyword">as</span> fin, <span class="built_in">open</span>(output_file, <span class="string">&#x27;w&#x27;</span>) <span class="keyword">as</span> fout:</span><br><span class="line">        <span class="keyword">for</span> line <span class="keyword">in</span> fin:</span><br><span class="line">            <span class="keyword">if</span> line.startswith(<span class="string">&#x27;&gt;&#x27;</span>):</span><br><span class="line">                seq_id = line[<span class="number">1</span>:].strip().split()[<span class="number">0</span>]</span><br><span class="line">                target_info = id_to_target.get(seq_id, <span class="literal">None</span>)</span><br><span class="line">                <span class="keyword">if</span> target_info:</span><br><span class="line">                    fout.write(<span class="string">f&#x27;&gt;<span class="subst">&#123;seq_id&#125;</span> Target=<span class="subst">&#123;target_info&#125;</span>\n&#x27;</span>)</span><br><span class="line">                <span class="keyword">else</span>:</span><br><span class="line">                    fout.write(line)</span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                fout.write(line)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">    gff_file = sys.argv[<span class="number">1</span>]</span><br><span class="line">    fasta_file = sys.argv[<span class="number">2</span>]</span><br><span class="line">    output_file = sys.argv[<span class="number">3</span>]</span><br><span class="line"></span><br><span class="line">    id_to_target = parse_gff_target(gff_file)</span><br><span class="line">    add_target_to_fasta(fasta_file, id_to_target, output_file)</span><br></pre></td></tr></table></figure>
<h2 id="get-compleasm-nt-seq-py"><a href="#get-compleasm-nt-seq-py" class="headerlink" title="get_compleasm_nt_seq.py"></a>get_compleasm_nt_seq.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># get_compleasm_nt_seq.py</span></span><br><span class="line"><span class="comment"># Usage：python get_compleasm_nt_seq.py &lt;nt_fasta_file&gt; &lt;aa_fasta_file&gt; &lt;output_file&gt;</span></span><br><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line"><span class="keyword">from</span> Bio.Seq <span class="keyword">import</span> Seq</span><br><span class="line"><span class="keyword">from</span> Bio.SeqRecord <span class="keyword">import</span> SeqRecord</span><br><span class="line"><span class="keyword">from</span> collections <span class="keyword">import</span> defaultdict</span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">extract_matching_nucleotides</span>(<span class="params">nt_fasta_file, aa_fasta_file, output_file</span>):</span><br><span class="line">    aa_dict = &#123;record.<span class="built_in">id</span>: record.seq <span class="keyword">for</span> record <span class="keyword">in</span> SeqIO.parse(aa_fasta_file, <span class="string">&quot;fasta&quot;</span>)&#125;</span><br><span class="line">    aa_ids = <span class="built_in">set</span>(aa_dict.keys())</span><br><span class="line"></span><br><span class="line">    nt_records = <span class="built_in">list</span>(SeqIO.parse(nt_fasta_file, <span class="string">&quot;fasta&quot;</span>))</span><br><span class="line">    match_map = defaultdict(<span class="built_in">list</span>)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> record <span class="keyword">in</span> nt_records:</span><br><span class="line">        desc = record.description</span><br><span class="line">        <span class="keyword">for</span> aa_id <span class="keyword">in</span> aa_ids:</span><br><span class="line">            <span class="keyword">if</span> <span class="string">f&quot;Target=<span class="subst">&#123;aa_id&#125;</span>&quot;</span> <span class="keyword">in</span> desc:</span><br><span class="line">                match_map[aa_id].append(record)</span><br><span class="line"></span><br><span class="line">    resolved_records = []</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> aa_id <span class="keyword">in</span> aa_ids:</span><br><span class="line">        matches = match_map.get(aa_id, [])</span><br><span class="line">        <span class="keyword">if</span> <span class="keyword">not</span> matches:</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line">        <span class="keyword">elif</span> <span class="built_in">len</span>(matches) == <span class="number">1</span>:</span><br><span class="line">            selected = matches[<span class="number">0</span>]</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">f&quot;WARNING: 蛋白质ID <span class="subst">&#123;aa_id&#125;</span> 匹配到多个核苷酸序列: <span class="subst">&#123;[r.<span class="built_in">id</span> <span class="keyword">for</span> r <span class="keyword">in</span> matches]&#125;</span>&quot;</span>)</span><br><span class="line">            selected, best_score = resolve_multiple_matches_by_translation(matches, aa_dict[aa_id], aa_id)</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">f&quot;→ 选中核苷酸序列: <span class="subst">&#123;selected.<span class="built_in">id</span>&#125;</span>（翻译相似度得分: <span class="subst">&#123;best_score&#125;</span>）&quot;</span>)</span><br><span class="line"></span><br><span class="line">        modified_record = SeqRecord(</span><br><span class="line">            seq=selected.seq,</span><br><span class="line">            <span class="built_in">id</span>=aa_id,</span><br><span class="line">            description=selected.description</span><br><span class="line">        )</span><br><span class="line">        resolved_records.append(modified_record)</span><br><span class="line"></span><br><span class="line">    SeqIO.write(resolved_records, output_file, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line"></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;\n蛋白质序列总数: <span class="subst">&#123;<span class="built_in">len</span>(aa_ids)&#125;</span>&quot;</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;核苷酸序列总数: <span class="subst">&#123;<span class="built_in">len</span>(nt_records)&#125;</span>&quot;</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;成功匹配并解析的核苷酸序列数: <span class="subst">&#123;<span class="built_in">len</span>(resolved_records)&#125;</span>&quot;</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">resolve_multiple_matches_by_translation</span>(<span class="params">nt_records, target_aa_seq, aa_id=<span class="literal">None</span></span>):</span><br><span class="line">    <span class="keyword">def</span> <span class="title function_">compare_seqs</span>(<span class="params">seq1, seq2</span>):</span><br><span class="line">        min_len = <span class="built_in">min</span>(<span class="built_in">len</span>(seq1), <span class="built_in">len</span>(seq2))</span><br><span class="line">        <span class="keyword">return</span> <span class="built_in">sum</span>(a == b <span class="keyword">for</span> a, b <span class="keyword">in</span> <span class="built_in">zip</span>(seq1[:min_len], seq2[:min_len]))</span><br><span class="line"></span><br><span class="line">    best_record = <span class="literal">None</span></span><br><span class="line">    best_score = -<span class="number">1</span></span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> record <span class="keyword">in</span> nt_records:</span><br><span class="line">        max_score = <span class="number">0</span></span><br><span class="line">        <span class="keyword">for</span> frame <span class="keyword">in</span> <span class="built_in">range</span>(<span class="number">3</span>):</span><br><span class="line">            translated = record.seq[frame:].translate(to_stop=<span class="literal">True</span>)</span><br><span class="line">            score = compare_seqs(translated, target_aa_seq)</span><br><span class="line">            <span class="keyword">if</span> score &gt; max_score:</span><br><span class="line">                max_score = score</span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> max_score &gt; best_score:</span><br><span class="line">            best_score = max_score</span><br><span class="line">            best_record = record</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> best_record, best_score</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">    nt_fasta_file = sys.argv[<span class="number">1</span>]</span><br><span class="line">    aa_fasta_file = sys.argv[<span class="number">2</span>]</span><br><span class="line">    output_file = sys.argv[<span class="number">3</span>]</span><br><span class="line"></span><br><span class="line">    extract_matching_nucleotides(nt_fasta_file, aa_fasta_file, output_file)</span><br></pre></td></tr></table></figure></article><div class="tag_share"><div class="post-share"><div class="social-share" data-image="../../../../img/touxiang.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><a class="pagination-related" href="../../../02/12/%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0%E7%BB%84/" title="比较基因组"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info"><div class="info-1"><div class="info-item-1">Previous</div><div class="info-item-2">比较基因组</div></div><div class="info-2"><div class="info-item-1">比较基因组各个步骤可能有多个方法都能达到相同目的，作为自学的，使用哪个方法是一个完全随缘的事情。 系统发育分析下载数据下载分析物种的fna基因组文件和gff注释文件。首先收集物种的GenBank号和物种名，如同下面一样保存在1个txt文档中，物种名和Genbank用制表符\t分隔: 123456GCA_037577465.1	Liposcelis_bostrychophilaGCA_042257085.1	Liposcelis_tricolorGCA_023512825.1	Liposcelis_brunneaGCA_950004255.1	Mesopsocus_fuscifronsGCA_963971405.1	Psococerastis_gibbosaGCA_964261705.1	Loensia_variegata 然后使用1个python脚本来自动下载这些物种的基因组和注释文件，程序运行依赖NCBI Datasets CLI 工具，失败了，就自动重新下载，最高重试8次。 123conda install ncbi-datasets-cli #...</div></div></div></a><a class="pagination-related" href="../../../05/05/hybpiper%E4%B8%AA%E4%BA%BA%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E/" title="hypiper个人使用说明"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info text-right"><div class="info-1"><div class="info-item-1">Next</div><div class="info-item-2">hypiper个人使用说明</div></div><div class="info-2"><div class="info-item-1">在进行一个类群的系统发育分析过程中，在利用Othofinder或是其他方法基于基因组或转录组数据得到一组单拷贝基因后，为了更快捷的获取其他所需物种的该组单拷贝基因，往往会考虑从基因组或转录组测序数据入手，因此会使用aTRAM或Hybpiper。在尝试使用aTRAM时没有成功，Hybpiper倒是跑通了。因此，在此只简要介绍个人使用Hypbpiper的记录，并且针对转录组数据。以前也有看到使用aTRAM从基因组低深度测序数据中，组装对应的单拷贝基因，再以后尝试吧。 0 软件安装sratoolkit安装1234cd bio_software/wget https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/3.2.1/sratoolkit.3.2.1-ubuntu64.tar.gztar zxvf sratoolkit.3.2.1-ubuntu64.tar.gz hybpiper安装12conda create -n hybpiper hybpiperconda activate hybpiper 1...</div></div></div></a></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info text-center"><div class="avatar-img"><img src="../../../../img/touxiang.jpg" onerror="this.onerror=null;this.src='../../../../img/friend_404.gif'" alt="avatar"/></div><div class="author-info-name">VERYXING565</div><div class="author-info-description">有事邮箱：veryxing565@163.com</div><div class="site-data"><a href="../../../../archives/"><div class="headline">Articles</div><div class="length-num">3</div></a><a href="../../../../tags/"><div class="headline">Tags</div><div class="length-num">0</div></a><a href="../../../../categories/"><div class="headline">Categories</div><div class="length-num">0</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/veryxing565"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Contents</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%A6%82%E8%BF%B0"><span class="toc-number">1.</span> <span class="toc-text">概述</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#1-%E4%B8%80%E8%BD%AE%E7%AD%9B%E9%80%89%EF%BC%9A%E5%9F%BA%E5%9B%A0%E7%BB%84"><span class="toc-number">2.</span> <span class="toc-text">1 一轮筛选：基因组</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#1-1-%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD"><span class="toc-number">2.1.</span> <span class="toc-text">1.1 数据下载</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#1-2-%E6%95%B0%E6%8D%AE%E7%AD%9B%E9%80%89%E5%92%8C%E6%8F%90%E5%8F%96%E5%8D%95%E6%8B%B7%E8%B4%9D%E5%9F%BA%E5%9B%A0"><span class="toc-number">2.2.</span> <span class="toc-text">1.2 数据筛选和提取单拷贝基因</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#1-3-%E6%95%B4%E7%90%86"><span class="toc-number">2.3.</span> <span class="toc-text">1.3 整理</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#2-%E4%BA%8C%E8%BD%AE%E7%AD%9B%E9%80%89%EF%BC%9A%E8%BD%AC%E5%BD%95%E7%BB%84"><span class="toc-number">3.</span> <span class="toc-text">2 二轮筛选：转录组</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%B5%81%E7%A8%8B%E4%B8%AD%E4%BD%BF%E7%94%A8%E7%9A%84%E8%84%9A%E6%9C%AC"><span class="toc-number">4.</span> <span class="toc-text">流程中使用的脚本</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#assembly-genome-download-link-py"><span class="toc-number">4.1.</span> <span class="toc-text">assembly_genome_download_link.py</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#compleasm-out-analyse-py"><span class="toc-number">4.2.</span> <span class="toc-text">compleasm_out_analyse.py</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#normal-gff-of-compleasm-py"><span class="toc-number">4.3.</span> <span class="toc-text">normal_gff_of_compleasm.py</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#add-description-py"><span class="toc-number">4.4.</span> <span class="toc-text">add_description.py</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#get-compleasm-nt-seq-py"><span class="toc-number">4.5.</span> <span class="toc-text">get_compleasm_nt_seq.py</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Posts</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="../../../05/05/hybpiper%E4%B8%AA%E4%BA%BA%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E/" title="hypiper个人使用说明">hypiper个人使用说明</a><time datetime="2025-05-05T06:16:46.000Z" title="Created 2025-05-05 14:16:46">2025-05-05</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="../../../03/18/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/" title="系统发育分析.md">系统发育分析.md</a><time datetime="2025-03-18T04:24:34.000Z" title="Created 2025-03-18 12:24:34">2025-03-18</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="" title="比较基因组">比较基因组</a><time datetime="2025-02-12T06:16:46.000Z" title="Created 2025-02-12 14:16:46">2025-02-12</time></div></div></div></div></div></div></main><footer id="footer" style="background-image: url(/img/footer.jpg);"><div id="footer-wrap"><div class="copyright">&copy;2025 By VERYXING565</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div><div class="footer_custom_text">Hi, welcome to my <a target="_blank" rel="noopener" href="https://butterfly.js.org/">blog</a>!</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Reading Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light and Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle Between Single-column and Double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="Settings"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back to Top"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="../../../../js/utils.js"></script><script src="../../../../js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/medium-zoom/dist/medium-zoom.min.js"></script><script src="https://cdn.jsdelivr.net/npm/node-snackbar/dist/snackbar.min.js"></script><div class="js-pjax"></div><canvas class="fireworks" mobile="false"></canvas><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/dist/fireworks.min.js"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>