<!DOCTYPE html><html lang="ZH" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>hypiper个人使用说明 | VEXYXING</title><meta name="author" content="VERYXING565"><meta name="copyright" content="VERYXING565"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="在进行一个类群的系统发育分析过程中，在利用Othofinder或是其他方法基于基因组或转录组数据得到一组单拷贝基因后，为了更快捷的获取其他所需物种的该组单拷贝基因，往往会考虑从基因组或转录组测序数据入手，因此会使用aTRAM或Hybpiper。在尝试使用aTRAM时没有成功，Hybpiper倒是跑通了。因此，在此只简要介绍个人使用Hypbpiper的记录，并且针对转录组数据。以前也有看到使用aTR">
<meta property="og:type" content="article">
<meta property="og:title" content="hypiper个人使用说明">
<meta property="og:url" content="http://example.com/2025/05/05/hybpiper%E4%B8%AA%E4%BA%BA%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E/index.html">
<meta property="og:site_name" content="VEXYXING">
<meta property="og:description" content="在进行一个类群的系统发育分析过程中，在利用Othofinder或是其他方法基于基因组或转录组数据得到一组单拷贝基因后，为了更快捷的获取其他所需物种的该组单拷贝基因，往往会考虑从基因组或转录组测序数据入手，因此会使用aTRAM或Hybpiper。在尝试使用aTRAM时没有成功，Hybpiper倒是跑通了。因此，在此只简要介绍个人使用Hypbpiper的记录，并且针对转录组数据。以前也有看到使用aTR">
<meta property="og:locale">
<meta property="og:image" content="http://example.com/img/touxiang.jpg">
<meta property="article:published_time" content="2025-05-05T06:16:46.000Z">
<meta property="article:modified_time" content="2025-05-05T15:36:47.085Z">
<meta property="article:author" content="VERYXING565">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/img/touxiang.jpg"><script type="application/ld+json">{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "hypiper个人使用说明",
  "url": "http://example.com/2025/05/05/hybpiper%E4%B8%AA%E4%BA%BA%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E/",
  "image": "http://example.com/img/touxiang.jpg",
  "datePublished": "2025-05-05T06:16:46.000Z",
  "dateModified": "2025-05-05T15:36:47.085Z",
  "author": [
    {
      "@type": "Person",
      "name": "VERYXING565",
      "url": "http://example.com/"
    }
  ]
}</script><link rel="shortcut icon" href="../../../../img/cat.jpg"><link rel="canonical" href="http://example.com/2025/05/05/hybpiper%E4%B8%AA%E4%BA%BA%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="../../../../css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/node-snackbar/dist/snackbar.min.css" media="print" onload="this.media='all'"><script>
    (() => {
      
    const saveToLocal = {
      set: (key, value, ttl) => {
        if (!ttl) return
        const expiry = Date.now() + ttl * 86400000
        localStorage.setItem(key, JSON.stringify({ value, expiry }))
      },
      get: key => {
        const itemStr = localStorage.getItem(key)
        if (!itemStr) return undefined
        const { value, expiry } = JSON.parse(itemStr)
        if (Date.now() > expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return value
      }
    }

    window.btf = {
      saveToLocal,
      getScript: (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        Object.entries(attr).forEach(([key, val]) => script.setAttribute(key, val))
        script.onload = script.onreadystatechange = () => {
          if (!script.readyState || /loaded|complete/.test(script.readyState)) resolve()
        }
        script.onerror = reject
        document.head.appendChild(script)
      }),
      getCSS: (url, id) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onload = link.onreadystatechange = () => {
          if (!link.readyState || /loaded|complete/.test(link.readyState)) resolve()
        }
        link.onerror = reject
        document.head.appendChild(link)
      }),
      addGlobalFn: (key, fn, name = false, parent = window) => {
        if (!false && key.startsWith('pjax')) return
        const globalFn = parent.globalFn || {}
        globalFn[key] = globalFn[key] || {}
        globalFn[key][name || Object.keys(globalFn[key]).length] = fn
        parent.globalFn = globalFn
      }
    }
  
      
      const activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      const activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }

      btf.activateDarkMode = activateDarkMode
      btf.activateLightMode = activateLightMode

      const theme = saveToLocal.get('theme')
    
          const mediaQueryDark = window.matchMedia('(prefers-color-scheme: dark)')
          const mediaQueryLight = window.matchMedia('(prefers-color-scheme: light)')
          
          if (theme === undefined) {
            if (mediaQueryLight.matches) activateLightMode()
            else if (mediaQueryDark.matches) activateDarkMode()
            else {
              const hour = new Date().getHours()
              const isNight = hour <= 6 || hour >= 18
              isNight ? activateDarkMode() : activateLightMode()
            }
            mediaQueryDark.addEventListener('change', () => {
              if (saveToLocal.get('theme') === undefined) {
                e.matches ? activateDarkMode() : activateLightMode()
              }
            })
          } else {
            theme === 'light' ? activateLightMode() : activateDarkMode()
          }
        
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        document.documentElement.classList.toggle('hide-aside', asideStatus === 'hide')
      }
    
      
    const detectApple = () => {
      if (/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)) {
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
  
    })()
  </script><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false,"highlightFullpage":false,"highlightMacStyle":false},
  copy: {
    success: 'Copy Successful',
    error: 'Copy Failed',
    noSupport: 'Browser Not Supported'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: 'Just now',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'medium_zoom',
  Snackbar: {"chs_to_cht":"You have switched to Traditional Chinese","cht_to_chs":"You have switched to Simplified Chinese","day_to_night":"You have switched to Dark Mode","night_to_day":"You have switched to Light Mode","bgLight":"#49b1f5","bgDark":"#2d3035","position":"top-center"},
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid/dist/infinitegrid.min.js',
    buttonText: 'Load More'
  },
  isPhotoFigcaption: true,
  islazyloadPlugin: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: true,
  },
  autoDarkmode: true
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: 'hypiper个人使用说明',
  isHighlightShrink: false,
  isToc: true,
  pageType: 'post'
}</script><meta name="generator" content="Hexo 7.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img text-center"><img src="../../../../img/touxiang.jpg" onerror="this.onerror=null;this.src='../../../../img/friend_404.gif'" alt="avatar"/></div><div class="site-data text-center"><a href="../../../../archives/"><div class="headline">Articles</div><div class="length-num">3</div></a><a href="../../../../tags/"><div class="headline">Tags</div><div class="length-num">0</div></a><a href="../../../../categories/"><div class="headline">Categories</div><div class="length-num">0</div></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="../../../../index.html"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="../../../../archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="../../../../comments/"><i class="fa-fw fas fa-envelope-open"></i><span> 留言板</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url(/img/top_image.png);"><nav id="nav"><span id="blog-info"><a class="nav-site-title" href="../../../../index.html"><span class="site-name">VEXYXING</span></a><a class="nav-page-title" href="../../../../index.html"><span class="site-name">hypiper个人使用说明</span></a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="../../../../index.html"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="../../../../archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="../../../../comments/"><i class="fa-fw fas fa-envelope-open"></i><span> 留言板</span></a></div></div><div id="toggle-menu"><span class="site-page"><i class="fas fa-bars fa-fw"></i></span></div></div></nav><div id="post-info"><h1 class="post-title">hypiper个人使用说明</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2025-05-05T06:16:46.000Z" title="Created 2025-05-05 14:16:46">2025-05-05</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2025-05-05T15:36:47.085Z" title="Updated 2025-05-05 23:36:47">2025-05-05</time></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title=""><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post Views:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="container post-content" id="article-container"><p>在进行一个类群的系统发育分析过程中，在利用Othofinder或是其他方法基于基因组或转录组数据得到一组单拷贝基因后，为了更快捷的获取其他所需物种的该组单拷贝基因，往往会考虑从基因组或转录组测序数据入手，因此会使用aTRAM或Hybpiper。在尝试使用aTRAM时没有成功，Hybpiper倒是跑通了。因此，在此只简要介绍个人使用Hypbpiper的记录，并且针对转录组数据。以前也有看到使用aTRAM从基因组低深度测序数据中，组装对应的单拷贝基因，再以后尝试吧。</p>
<h1 id="1-数据选择"><a href="#1-数据选择" class="headerlink" title="1 数据选择"></a>1 数据选择</h1><p>转录组的数据主要源自NCBI，通过搜索一个分类群，然后挑选物种及其转录组数据。</p>
<p>若一个物种的RNA-seq数据数量较多，则优先选择该物种，并且尽可能下载该物种的多个RNA-seq数据，主要利用sratoolkit。如在昆虫纲的Grylloblattodea中，而Grylloblattella pravdini的转录组有9个，数量最多，因此倾向于选择该物种，下载其尽可能多的转录组数据。而本文假设下载SRR11533876、SRR11533877	和SRR11533888三个（实际上认为应该更多一些）。</p>
<p><img src="/2025/05/05/hybpiper%E4%B8%AA%E4%BA%BA%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E/%E5%9B%BE1.png"></p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> workdir <span class="comment"># 创建工作目录</span></span><br><span class="line"><span class="built_in">cd</span> workdir <span class="comment"># 进入工作目录</span></span><br><span class="line">path/to/sratoolkit/bin/prefetch SRR11533876 SRR11533877 SRR11533888 <span class="comment"># 下载3个转录组数据</span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> SRR*</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">path/to/sratoolkit/bin/fastq-dump --split-3 <span class="variable">$i</span>/<span class="variable">$i</span>.sra <span class="comment"># 批量拆分双端测序文件</span></span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<h1 id="2-质控和过滤"><a href="#2-质控和过滤" class="headerlink" title="2 质控和过滤"></a>2 质控和过滤</h1><p>使用fastp进行质控</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> SRR11533876</span><br><span class="line">fastp -i SRR11533876_1.fastq -I SRR11533876_2.fastq -o SRR11533876_1.clean.fastq -O SRR11533876_2.clean.fastq</span><br><span class="line"><span class="built_in">cd</span> ../SRR11533877</span><br><span class="line">fastp -i SRR11533877_1.fastq -I SRR11533877_2.fastq -o SRR11533877_1.clean.fastq -O SRR11533877_2.clean.fastq</span><br><span class="line"><span class="built_in">cd</span> ../SRR11533888</span><br><span class="line">fastp -i SRR11533888_1.fastq -I SRR11533888_2.fastq -o SRR11533888_1.clean.fastq -O SRR11533888_2.clean.fastq</span><br><span class="line"><span class="built_in">cd</span> ..</span><br></pre></td></tr></table></figure>
<h1 id="3-Hybpiper运行"><a href="#3-Hybpiper运行" class="headerlink" title="3 Hybpiper运行"></a>3 Hybpiper运行</h1><p>Hybpiper的运行需要近缘物种的确定的单拷贝基因序列，我们采用蛋白质序列，并且测序数据的格式为fastq，因此，比对软件使用blast。而Hypiper的运行可以对单个转录组分别进行，也可以将多个转录组数据合并在一起，再使用Hybpiper组装单拷贝基因。在实际测试中，合并后的转录组组装效果更好一些。</p>
<h2 id="3-1-单个转录组分别组装"><a href="#3-1-单个转录组分别组装" class="headerlink" title="3.1 单个转录组分别组装"></a>3.1 单个转录组分别组装</h2><p>单个转录组的单拷贝基因组装，使用blast。然后通过python汇集多个转录组的组装结果，并剔除序列长度和参考序列差异过大的，以及N占比过高的序列，若还有多个序列候选，则选择与参考序列最相似的序列，相似性的分析需以来mafft。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> SSR*</span><br><span class="line"><span class="keyword">do</span> </span><br><span class="line"><span class="built_in">cd</span> <span class="variable">$i</span></span><br><span class="line">hybpiper assemble -r <span class="variable">$i_1</span>.clean.fastq <span class="variable">$i_2</span>.clean.fastq -t_aa path/to/ref_aa.fa –-cpu 20 <span class="comment"># 从一个转录组中组装单拷贝基因</span></span><br><span class="line"><span class="built_in">cd</span> ..</span><br><span class="line"><span class="keyword">done</span></span><br><span class="line">python muti-hybpiper_out_analyse.py ./ Grylloblattella_pravdini ./</span><br><span class="line"><span class="comment"># 将多个转录组的组装结果汇集</span></span><br><span class="line">python filter_hybpiper_fasta.py Grylloblattella_pravdini_gather.fa path/to/ref_aa.fa Grylloblattella_pravdini_out.fa 0.15 0.15</span><br><span class="line"><span class="comment"># 筛选多个转录组的组装汇集结果</span></span><br></pre></td></tr></table></figure>
<h2 id="3-2-转录组先合并再组装"><a href="#3-2-转录组先合并再组装" class="headerlink" title="3.2 转录组先合并再组装"></a>3.2 转录组先合并再组装</h2><p>即先将多个转录组合并，在利用hybpiper组装单拷贝基因。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">mkdir</span> merge</span><br><span class="line"><span class="built_in">cd</span> merge</span><br><span class="line"><span class="built_in">cat</span> ../SRR11533876/SRR11533876_1.clean.fastq ../SRR11533877/SRR11533877_1.clean.fastq ../SRR11533888/SRR11533888_1.clean.fastq &gt; merge_1.clean.fastq</span><br><span class="line"><span class="built_in">cat</span> ../SRR11533876/SRR11533876_2.clean.fastq ../SRR11533877/SRR11533877_2.clean.fastq ../SRR11533888/SRR11533888_2.clean.fastq &gt; merge_2.clean.fastq</span><br><span class="line"><span class="comment"># 合并多个转录组的两端测序的文件</span></span><br><span class="line">hybpiper assemble -r merge_1.clean.fastq merge_2.clean.fastq -t_aa path/to/ref_aa.fa –-cpu 20 </span><br><span class="line"><span class="comment"># 组装单拷贝基因</span></span><br><span class="line">python muti-hybpiper_out_analyse.py ./ Grylloblattella_pravdini ./</span><br><span class="line"><span class="comment"># 整理组装结果</span></span><br><span class="line">python filter_hybpiper_fasta.py Grylloblattella_pravdini_gather.fa path/to/ref_aa.fa Grylloblattella_pravdini_out.fa 0.15 0.15</span><br><span class="line"><span class="comment"># 过滤组装结果</span></span><br></pre></td></tr></table></figure>
<h2 id="3-3-利用diamond加速组装"><a href="#3-3-利用diamond加速组装" class="headerlink" title="3.3 利用diamond加速组装"></a>3.3 利用diamond加速组装</h2><p>利用diamond完成在hybpiper组装中的比对工作，速度比默认的blast快很多。但是diamond不支持fastq格式的数据，需要先将fastq转为fasta。但没有试过，步骤基本如下</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">conda install seqtk -c bioconda <span class="comment"># 安装seqtk</span></span><br><span class="line">seqtk <span class="built_in">seq</span> -a merge_1.clean.fastq &gt; merge_1.clean.fasta</span><br><span class="line">seqtk <span class="built_in">seq</span> -a merge_2.clean.fastq &gt; merge_2.clean.fasta</span><br><span class="line"><span class="comment"># fastq转fasta</span></span><br><span class="line">hybpiper assemble -r SRR921624_1.fasta SRR921624_2.fasta -t_aa path/to/ref_aa.fa --diamond --cpu 20</span><br><span class="line">python muti-hybpiper_out_analyse.py ./ Grylloblattella_pravdini ./</span><br><span class="line"><span class="comment"># 整理组装结果</span></span><br><span class="line">python filter_hybpiper_fasta.py Grylloblattella_pravdini_gather.fa path/to/ref_aa.fa Grylloblattella_pravdini_out.fa 0.15 0.15</span><br><span class="line"><span class="comment"># 过滤组装结果</span></span><br></pre></td></tr></table></figure>
<h1 id="4-Python脚本"><a href="#4-Python脚本" class="headerlink" title="4 Python脚本"></a>4 Python脚本</h1><h2 id="4-1-muti-hybpiper-out-analyse-py"><a href="#4-1-muti-hybpiper-out-analyse-py" class="headerlink" title="4.1 muti-hybpiper_out_analyse.py"></a>4.1 muti-hybpiper_out_analyse.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># muti-hybpiper_out_analyse.py</span></span><br><span class="line"><span class="comment"># Usage: python muti-hybpiper_out_analyse.py &lt;workdir&gt; &lt;Species_name&gt; &lt;output_dir&gt;</span></span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">main</span>(<span class="params">path2work, species_name, statistical_result</span>):</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;正在扫描主目录: <span class="subst">&#123;path2work&#125;</span>&quot;</span>)</span><br><span class="line">    folders = [os.path.join(path2work, name) <span class="keyword">for</span> name <span class="keyword">in</span> os.listdir(path2work) <span class="keyword">if</span></span><br><span class="line">               os.path.isdir(os.path.join(path2work, name))]</span><br><span class="line"></span><br><span class="line">    <span class="keyword">if</span> <span class="keyword">not</span> folders:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">&quot;未找到任何HypiPiper结果文件夹。请确认路径正确。&quot;</span>)</span><br><span class="line">        <span class="keyword">return</span></span><br><span class="line"></span><br><span class="line">    all_records = []</span><br><span class="line">    <span class="keyword">for</span> folder <span class="keyword">in</span> folders:</span><br><span class="line">        hypiper_foler_name = os.path.basename(folder)</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">f&quot;\n正在处理：<span class="subst">&#123;hypiper_foler_name&#125;</span>&quot;</span>)</span><br><span class="line">        records = hypiper_analyses(folder, species_name, statistical_result)</span><br><span class="line">        all_records.extend(records)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 写入所有序列为一个FASTA文件</span></span><br><span class="line">    gather_fasta = os.path.join(statistical_result, <span class="string">f&quot;<span class="subst">&#123;species_name&#125;</span>_gather.fa&quot;</span>)</span><br><span class="line">    SeqIO.write(all_records, gather_fasta, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;\n所有序列已汇总并保存至: <span class="subst">&#123;gather_fasta&#125;</span>&quot;</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;共收集序列数: <span class="subst">&#123;<span class="built_in">len</span>(all_records)&#125;</span> 条&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">hypiper_analyses</span>(<span class="params">folder, species_name, statistical_result</span>):</span><br><span class="line">    results = []</span><br><span class="line">    all_records = []</span><br><span class="line">    hypiper_foler_name = os.path.basename(folder)</span><br><span class="line">    path2hypiper = os.path.join(folder, hypiper_foler_name)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">if</span> <span class="keyword">not</span> os.path.exists(path2hypiper):</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">f&quot;跳过: <span class="subst">&#123;path2hypiper&#125;</span> 不存在&quot;</span>)</span><br><span class="line">        <span class="keyword">return</span> []</span><br><span class="line"></span><br><span class="line">    gene_folders = [os.path.join(path2hypiper, name) <span class="keyword">for</span> name <span class="keyword">in</span> os.listdir(path2hypiper) <span class="keyword">if</span></span><br><span class="line">                    os.path.isdir(os.path.join(path2hypiper, name))]</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> gene_folder <span class="keyword">in</span> gene_folders:</span><br><span class="line">        gene_foler_name = os.path.basename(gene_folder)</span><br><span class="line">        path2seq = os.path.join(gene_folder, hypiper_foler_name, <span class="string">&quot;sequences&quot;</span>, <span class="string">&quot;FAA&quot;</span>, <span class="string">f&quot;<span class="subst">&#123;gene_foler_name&#125;</span>.FAA&quot;</span>)</span><br><span class="line">        path2warming = os.path.join(gene_folder, hypiper_foler_name, <span class="string">&quot;paralog_warning_by_contig_depth.txt&quot;</span>)</span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> <span class="keyword">not</span> os.path.exists(path2seq):</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">f&quot;未找到序列文件: <span class="subst">&#123;path2seq&#125;</span>，跳过该基因&quot;</span>)</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line"></span><br><span class="line">        judge = <span class="number">0</span></span><br><span class="line">        <span class="keyword">if</span> os.path.exists(path2warming):</span><br><span class="line">            judge = <span class="number">1</span></span><br><span class="line"></span><br><span class="line">        <span class="keyword">try</span>:</span><br><span class="line">            records = <span class="built_in">list</span>(SeqIO.parse(path2seq, <span class="string">&quot;fasta&quot;</span>))</span><br><span class="line">        <span class="keyword">except</span> Exception <span class="keyword">as</span> e:</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">f&quot;解析失败：<span class="subst">&#123;path2seq&#125;</span> 错误信息: <span class="subst">&#123;e&#125;</span>&quot;</span>)</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line"></span><br><span class="line">        Number_of_Seq = <span class="built_in">len</span>(records)</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">f&quot;  - <span class="subst">&#123;gene_foler_name&#125;</span>：共 <span class="subst">&#123;Number_of_Seq&#125;</span> 条序列，Warming=<span class="subst">&#123;judge&#125;</span>&quot;</span>)</span><br><span class="line"></span><br><span class="line">        <span class="keyword">for</span> record <span class="keyword">in</span> records:</span><br><span class="line">            sequence = <span class="built_in">str</span>(record.seq).upper()</span><br><span class="line">            seq_length = <span class="built_in">len</span>(sequence)</span><br><span class="line">            x_count = sequence.count(<span class="string">&quot;X&quot;</span>)</span><br><span class="line">            x_ratio = x_count / seq_length <span class="keyword">if</span> seq_length &gt; <span class="number">0</span> <span class="keyword">else</span> <span class="number">0</span></span><br><span class="line">            record.<span class="built_in">id</span> = <span class="string">f&quot;<span class="subst">&#123;species_name&#125;</span>_<span class="subst">&#123;hypiper_foler_name&#125;</span>-<span class="subst">&#123;gene_foler_name&#125;</span>&quot;</span></span><br><span class="line">            record.description = <span class="string">f&quot;seq_length:<span class="subst">&#123;seq_length&#125;</span> x_count:<span class="subst">&#123;x_count&#125;</span> x_ratio:<span class="subst">&#123;x_ratio&#125;</span> warming:<span class="subst">&#123;judge&#125;</span>&quot;</span></span><br><span class="line">            all_records.append(record)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> all_records</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">    <span class="keyword">if</span> <span class="built_in">len</span>(sys.argv) != <span class="number">4</span>:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">&quot;Usage: python muti-hybpiper_out_analyse.py &lt;workdir&gt; &lt;Species_name&gt; &lt;output_dir&gt;&quot;</span>)</span><br><span class="line">        sys.exit(<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line">    path2work = sys.argv[<span class="number">1</span>]</span><br><span class="line">    species_name = sys.argv[<span class="number">2</span>]</span><br><span class="line">    statistical_result = sys.argv[<span class="number">3</span>]</span><br><span class="line"></span><br><span class="line">    main(path2work, species_name, statistical_result)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h2 id="4-2-filter-hybpiper-fasta-py"><a href="#4-2-filter-hybpiper-fasta-py" class="headerlink" title="4.2 filter_hybpiper_fasta.py"></a>4.2 filter_hybpiper_fasta.py</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># filter_hybpiper_fasta.py</span></span><br><span class="line"><span class="comment"># Usage: python filter_hybpiper_fasta.py &lt;input_fasta&gt; &lt;ref_fasta&gt; &lt;output_fasta&gt; [max_x_ratio] [max_length_diff_ratio]</span></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> subprocess</span><br><span class="line"><span class="keyword">from</span> Bio <span class="keyword">import</span> SeqIO</span><br><span class="line"><span class="keyword">from</span> tempfile <span class="keyword">import</span> mkdtemp</span><br><span class="line"><span class="keyword">from</span> shutil <span class="keyword">import</span> rmtree</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">run_mafft</span>(<span class="params">seq1, seq2</span>):</span><br><span class="line">    temp_dir = mkdtemp()</span><br><span class="line">    file1 = os.path.join(temp_dir, <span class="string">&quot;seq1.fasta&quot;</span>)</span><br><span class="line">    file2 = os.path.join(temp_dir, <span class="string">&quot;seq2.fasta&quot;</span>)</span><br><span class="line">    combined = os.path.join(temp_dir, <span class="string">&quot;combined.fasta&quot;</span>)</span><br><span class="line">    aligned = os.path.join(temp_dir, <span class="string">&quot;aligned.fasta&quot;</span>)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(combined, <span class="string">&#x27;w&#x27;</span>) <span class="keyword">as</span> out:</span><br><span class="line">        out.write(<span class="string">f&quot;&gt;seq1\n<span class="subst">&#123;seq1&#125;</span>\n&gt;seq2\n<span class="subst">&#123;seq2&#125;</span>\n&quot;</span>)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">try</span>:</span><br><span class="line">        subprocess.run([<span class="string">&quot;mafft&quot;</span>, <span class="string">&quot;--quiet&quot;</span>, combined], stdout=<span class="built_in">open</span>(aligned, <span class="string">&#x27;w&#x27;</span>), check=<span class="literal">True</span>)</span><br><span class="line">        records = <span class="built_in">list</span>(SeqIO.parse(aligned, <span class="string">&quot;fasta&quot;</span>))</span><br><span class="line">        aln1, aln2 = <span class="built_in">str</span>(records[<span class="number">0</span>].seq), <span class="built_in">str</span>(records[<span class="number">1</span>].seq)</span><br><span class="line">        <span class="keyword">match</span> = <span class="built_in">sum</span>(a == b <span class="keyword">for</span> a, b <span class="keyword">in</span> <span class="built_in">zip</span>(aln1, aln2) <span class="keyword">if</span> a != <span class="string">&#x27;-&#x27;</span> <span class="keyword">and</span> b != <span class="string">&#x27;-&#x27;</span>)</span><br><span class="line">        length = <span class="built_in">sum</span>(<span class="number">1</span> <span class="keyword">for</span> a, b <span class="keyword">in</span> <span class="built_in">zip</span>(aln1, aln2) <span class="keyword">if</span> a != <span class="string">&#x27;-&#x27;</span> <span class="keyword">and</span> b != <span class="string">&#x27;-&#x27;</span>)</span><br><span class="line">        similarity = <span class="keyword">match</span> / length <span class="keyword">if</span> length <span class="keyword">else</span> <span class="number">0</span></span><br><span class="line">        <span class="keyword">return</span> similarity</span><br><span class="line">    <span class="keyword">except</span> Exception <span class="keyword">as</span> e:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">f&quot;MAFFT比对失败: <span class="subst">&#123;e&#125;</span>&quot;</span>)</span><br><span class="line">        <span class="keyword">return</span> <span class="number">0</span></span><br><span class="line">    <span class="keyword">finally</span>:</span><br><span class="line">        rmtree(temp_dir)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">parse_record_description</span>(<span class="params">record</span>):</span><br><span class="line">    desc = record.description</span><br><span class="line">    info = &#123;<span class="string">&quot;seq_length&quot;</span>: <span class="built_in">len</span>(record.seq), <span class="string">&quot;x_count&quot;</span>: record.seq.upper().count(<span class="string">&quot;X&quot;</span>), <span class="string">&quot;x_ratio&quot;</span>: <span class="number">0.0</span>, <span class="string">&quot;warming&quot;</span>: <span class="number">1</span>&#125;</span><br><span class="line">    <span class="keyword">try</span>:</span><br><span class="line">        <span class="keyword">for</span> part <span class="keyword">in</span> desc.split():</span><br><span class="line">            <span class="keyword">if</span> part.startswith(<span class="string">&quot;seq_length:&quot;</span>):</span><br><span class="line">                info[<span class="string">&quot;seq_length&quot;</span>] = <span class="built_in">int</span>(part.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>])</span><br><span class="line">            <span class="keyword">elif</span> part.startswith(<span class="string">&quot;x_count:&quot;</span>):</span><br><span class="line">                info[<span class="string">&quot;x_count&quot;</span>] = <span class="built_in">int</span>(part.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>])</span><br><span class="line">            <span class="keyword">elif</span> part.startswith(<span class="string">&quot;x_ratio:&quot;</span>):</span><br><span class="line">                info[<span class="string">&quot;x_ratio&quot;</span>] = <span class="built_in">float</span>(part.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>])</span><br><span class="line">            <span class="keyword">elif</span> part.startswith(<span class="string">&quot;warming:&quot;</span>):</span><br><span class="line">                info[<span class="string">&quot;warming&quot;</span>] = <span class="built_in">int</span>(part.split(<span class="string">&quot;:&quot;</span>)[<span class="number">1</span>])</span><br><span class="line">    <span class="keyword">except</span>:</span><br><span class="line">        <span class="keyword">pass</span></span><br><span class="line">    <span class="keyword">return</span> info</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">filter_best_sequences</span>(<span class="params">gene_to_records, reference_records, max_n_ratio=<span class="number">0.5</span>, max_length_diff_ratio=<span class="number">0.2</span></span>):</span><br><span class="line">    retained = []</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> gene, records <span class="keyword">in</span> gene_to_records.items():</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">f&quot;\n正在处理基因: <span class="subst">&#123;gene&#125;</span>（候选序列数: <span class="subst">&#123;<span class="built_in">len</span>(records)&#125;</span>）&quot;</span>)</span><br><span class="line">        ref_seq = reference_records.get(gene, <span class="literal">None</span>)</span><br><span class="line"></span><br><span class="line">        qualified = []</span><br><span class="line">        <span class="keyword">for</span> r <span class="keyword">in</span> records:</span><br><span class="line">            info = parse_record_description(r)</span><br><span class="line">            r.annotations = info</span><br><span class="line">            <span class="keyword">if</span> info[<span class="string">&quot;x_ratio&quot;</span>] &gt; max_n_ratio:</span><br><span class="line">                <span class="built_in">print</span>(<span class="string">f&quot;X占比过高 (<span class="subst">&#123;info[<span class="string">&#x27;x_ratio&#x27;</span>]:<span class="number">.2</span>f&#125;</span>)，剔除: <span class="subst">&#123;r.<span class="built_in">id</span>&#125;</span>&quot;</span>)</span><br><span class="line">                <span class="keyword">continue</span></span><br><span class="line">            <span class="keyword">if</span> ref_seq:</span><br><span class="line">                ref_len = <span class="built_in">len</span>(ref_seq.seq)</span><br><span class="line">                len_diff_ratio = <span class="built_in">abs</span>(<span class="built_in">len</span>(r.seq) - ref_len) / ref_len</span><br><span class="line">                <span class="keyword">if</span> len_diff_ratio &gt; max_length_diff_ratio:</span><br><span class="line">                    <span class="built_in">print</span>(<span class="string">f&quot;长度差异过大 (<span class="subst">&#123;len_diff_ratio:<span class="number">.2</span>%&#125;</span>)，剔除: <span class="subst">&#123;r.<span class="built_in">id</span>&#125;</span>&quot;</span>)</span><br><span class="line">                    <span class="keyword">continue</span></span><br><span class="line">            qualified.append(r)</span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> <span class="keyword">not</span> qualified:</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">&quot;无合格序列，全部剔除&quot;</span>)</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> <span class="built_in">len</span>(qualified) == <span class="number">1</span>:</span><br><span class="line">            selected = qualified[<span class="number">0</span>]</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">f&quot;唯一合格序列保留: <span class="subst">&#123;selected.<span class="built_in">id</span>&#125;</span>&quot;</span>)</span><br><span class="line">            retained.append(selected)</span><br><span class="line">            <span class="keyword">continue</span></span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> ref_seq:</span><br><span class="line">            selected = <span class="built_in">max</span>(qualified, key=<span class="keyword">lambda</span> r: run_mafft(<span class="built_in">str</span>(r.seq), <span class="built_in">str</span>(ref_seq.seq)))</span><br><span class="line">            similarity = run_mafft(<span class="built_in">str</span>(selected.seq), <span class="built_in">str</span>(ref_seq.seq))</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">f&quot;保留与参考序列相似度最高者: <span class="subst">&#123;selected.<span class="built_in">id</span>&#125;</span>（相似度: <span class="subst">&#123;similarity:<span class="number">.2</span>%&#125;</span>）&quot;</span>)</span><br><span class="line">            retained.append(selected)</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            selected = qualified[<span class="number">0</span>]</span><br><span class="line">            <span class="built_in">print</span>(<span class="string">f&quot;无参考序列，随机保留第一个合格序列: <span class="subst">&#123;selected.<span class="built_in">id</span>&#125;</span>&quot;</span>)</span><br><span class="line">            retained.append(selected)</span><br><span class="line"></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;\n共保留 <span class="subst">&#123;<span class="built_in">len</span>(retained)&#125;</span> 条序列&quot;</span>)</span><br><span class="line">    <span class="keyword">for</span> r <span class="keyword">in</span> retained:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">f&quot;   保留序列: <span class="subst">&#123;r.<span class="built_in">id</span>&#125;</span>&quot;</span>)</span><br><span class="line">    <span class="keyword">return</span> retained</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">read_reference_fasta</span>(<span class="params">reference_fasta</span>):</span><br><span class="line">    ref_records = &#123;&#125;</span><br><span class="line">    <span class="keyword">for</span> record <span class="keyword">in</span> SeqIO.parse(reference_fasta, <span class="string">&quot;fasta&quot;</span>):</span><br><span class="line">        gene = record.<span class="built_in">id</span>.split(<span class="string">&quot;-&quot;</span>)[-<span class="number">1</span>]</span><br><span class="line">        ref_records[gene] = record</span><br><span class="line">    <span class="keyword">return</span> ref_records</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">group_sequences_by_gene</span>(<span class="params">input_fasta</span>):</span><br><span class="line">    gene_to_records = &#123;&#125;</span><br><span class="line">    <span class="keyword">for</span> record <span class="keyword">in</span> SeqIO.parse(input_fasta, <span class="string">&quot;fasta&quot;</span>):</span><br><span class="line">        gene = record.<span class="built_in">id</span>.split(<span class="string">&quot;-&quot;</span>)[-<span class="number">1</span>]</span><br><span class="line">        gene_to_records.setdefault(gene, []).append(record)</span><br><span class="line">    <span class="keyword">return</span> gene_to_records</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">write_fasta</span>(<span class="params">records, output_file</span>):</span><br><span class="line">    SeqIO.write(records, output_file, <span class="string">&quot;fasta&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">main</span>():</span><br><span class="line">    <span class="keyword">if</span> <span class="built_in">len</span>(sys.argv) &lt; <span class="number">4</span>:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">&quot;Usage: python filter_hybpiper_fasta.py &lt;input_fasta&gt; &lt;ref_fasta&gt; &lt;output_fasta&gt; [max_x_ratio] [max_length_diff_ratio]]&quot;</span>)</span><br><span class="line">        sys.exit(<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line">    input_fasta = sys.argv[<span class="number">1</span>]</span><br><span class="line">    reference_fasta = sys.argv[<span class="number">2</span>]</span><br><span class="line">    output_fasta = sys.argv[<span class="number">3</span>]</span><br><span class="line">    max_n_ratio = <span class="built_in">float</span>(sys.argv[<span class="number">4</span>]) <span class="keyword">if</span> <span class="built_in">len</span>(sys.argv) &gt; <span class="number">4</span> <span class="keyword">else</span> <span class="number">0.5</span></span><br><span class="line">    max_length_diff_ratio = <span class="built_in">float</span>(sys.argv[<span class="number">5</span>]) <span class="keyword">if</span> <span class="built_in">len</span>(sys.argv) &gt; <span class="number">5</span> <span class="keyword">else</span> <span class="number">0.2</span></span><br><span class="line"></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&quot;开始读取参考序列...&quot;</span>)</span><br><span class="line">    reference_records = read_reference_fasta(reference_fasta)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;读取参考序列完成，共 <span class="subst">&#123;<span class="built_in">len</span>(reference_records)&#125;</span> 条&quot;</span>)</span><br><span class="line"></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&quot;开始读取输入fasta...&quot;</span>)</span><br><span class="line">    gene_to_records = group_sequences_by_gene(input_fasta)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;读取完成，共涉及 <span class="subst">&#123;<span class="built_in">len</span>(gene_to_records)&#125;</span> 个基因&quot;</span>)</span><br><span class="line"></span><br><span class="line">    retained = filter_best_sequences(gene_to_records, reference_records, max_n_ratio, max_length_diff_ratio)</span><br><span class="line"></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f&quot;写入筛选结果到 <span class="subst">&#123;output_fasta&#125;</span> ...&quot;</span>)</span><br><span class="line">    write_fasta(retained, output_fasta)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&quot;完成&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">    main()</span><br><span class="line"></span><br></pre></td></tr></table></figure></article><div class="tag_share"><div class="post-share"><div class="social-share" data-image="../../../../img/touxiang.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><a class="pagination-related full-width" href="../../../03/18/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/" title="系统发育分析.md"><div class="cover" style="background: var(--default-bg-color)"></div><div class="info"><div class="info-1"><div class="info-item-1">Previous</div><div class="info-item-2">系统发育分析.md</div></div><div class="info-2"><div class="info-item-1">概述为了寻求各物种之间的系统发育关系，而去寻找各物种之间共有的同源单拷贝基因。我认为难点在于筛选数据，目前指定筛选的顺序为：  基因组：在一个目标分类单位中，如螯肢动物亚门，寻找其中所有可用的基因组数据，使用compleasm进行检验，保证使用的基因组具备高完整性和低冗余性，从而有较多的同源单拷贝基因可供选择。这里的阈值（S）我选择90%。即为第一轮筛选。  此时，得到一批基因组数据及其一组同源单拷贝基因，同时，提取一个物种的这组同源单拷贝基因作为参考。  基因组二代测序数据：这是第二轮数据筛选，从NCBI上下载缺失数据的分类单位的二代测序数据，然后使用参考的单拷贝基因数据进行搜索该物种对应的单拷贝基因数据，软件使用的是aTRAM。  转录组测序数据：这是第三轮数据筛选，即使用Trinity组装转录本，提取单拷贝基因   1 一轮筛选：基因组1.1...</div></div></div></a></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info text-center"><div class="avatar-img"><img src="../../../../img/touxiang.jpg" onerror="this.onerror=null;this.src='../../../../img/friend_404.gif'" alt="avatar"/></div><div class="author-info-name">VERYXING565</div><div class="author-info-description">有事邮箱：veryxing565@163.com</div><div class="site-data"><a href="../../../../archives/"><div class="headline">Articles</div><div class="length-num">3</div></a><a href="../../../../tags/"><div class="headline">Tags</div><div class="length-num">0</div></a><a href="../../../../categories/"><div class="headline">Categories</div><div class="length-num">0</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/veryxing565"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Contents</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#1-%E6%95%B0%E6%8D%AE%E9%80%89%E6%8B%A9"><span class="toc-number">1.</span> <span class="toc-text">1 数据选择</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#2-%E8%B4%A8%E6%8E%A7%E5%92%8C%E8%BF%87%E6%BB%A4"><span class="toc-number">2.</span> <span class="toc-text">2 质控和过滤</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#3-Hybpiper%E8%BF%90%E8%A1%8C"><span class="toc-number">3.</span> <span class="toc-text">3 Hybpiper运行</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#3-1-%E5%8D%95%E4%B8%AA%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E5%88%AB%E7%BB%84%E8%A3%85"><span class="toc-number">3.1.</span> <span class="toc-text">3.1 单个转录组分别组装</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#3-2-%E8%BD%AC%E5%BD%95%E7%BB%84%E5%85%88%E5%90%88%E5%B9%B6%E5%86%8D%E7%BB%84%E8%A3%85"><span class="toc-number">3.2.</span> <span class="toc-text">3.2 转录组先合并再组装</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#3-3-%E5%88%A9%E7%94%A8diamond%E5%8A%A0%E9%80%9F%E7%BB%84%E8%A3%85"><span class="toc-number">3.3.</span> <span class="toc-text">3.3 利用diamond加速组装</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#4-Python%E8%84%9A%E6%9C%AC"><span class="toc-number">4.</span> <span class="toc-text">4 Python脚本</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#4-1-muti-hybpiper-out-analyse-py"><span class="toc-number">4.1.</span> <span class="toc-text">4.1 muti-hybpiper_out_analyse.py</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#4-2-filter-hybpiper-fasta-py"><span class="toc-number">4.2.</span> <span class="toc-text">4.2 filter_hybpiper_fasta.py</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Posts</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="../../../05/05/hybpiper%E4%B8%AA%E4%BA%BA%E4%BD%BF%E7%94%A8%E8%AF%B4%E6%98%8E/" title="hypiper个人使用说明">hypiper个人使用说明</a><time datetime="2025-05-05T06:16:46.000Z" title="Created 2025-05-05 14:16:46">2025-05-05</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="../../../03/18/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/" title="系统发育分析.md">系统发育分析.md</a><time datetime="2025-03-18T04:24:34.000Z" title="Created 2025-03-18 12:24:34">2025-03-18</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="" title="比较基因组">比较基因组</a><time datetime="2025-02-12T06:16:46.000Z" title="Created 2025-02-12 14:16:46">2025-02-12</time></div></div></div></div></div></div></main><footer id="footer" style="background-image: url(/img/footer.jpg);"><div id="footer-wrap"><div class="copyright">&copy;2025 By VERYXING565</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div><div class="footer_custom_text">Hi, welcome to my <a target="_blank" rel="noopener" href="https://butterfly.js.org/">blog</a>!</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Reading Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light and Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle Between Single-column and Double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="Settings"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back to Top"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="../../../../js/utils.js"></script><script src="../../../../js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/medium-zoom/dist/medium-zoom.min.js"></script><script src="https://cdn.jsdelivr.net/npm/node-snackbar/dist/snackbar.min.js"></script><div class="js-pjax"></div><canvas class="fireworks" mobile="false"></canvas><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/dist/fireworks.min.js"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>